### ISSUE TITLE
[BUG?] Generated QR Codes have different padding

### ISSUE THREAD

**Issue by temuri416** [Bacon/BaconQrCode#26]
Hi there,

Thanks for the great library!

The code:

```php
$a = '{"i":"xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx","v":1,"t":1,"d":"9ZGeRh5hsoYuSW8uT0NZyXMa17FOPTa28Ykt8th9\/XP7PGNgcxXUbZwtcPB\/5hql","ud":"--","un":105}';
$renderer = new \BaconQrCode\Renderer\Image\Png();
$renderer->setHeight(210);
$renderer->setWidth(210);
$renderer->setMargin(0);
$writer = new \BaconQrCode\Writer($renderer);
$writer->writeFile($a, 'qrcode1.png', 'UTF-8', 2);

$b = '{"i":"xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx","v":1,"t":1,"d":"\/x6lV8xzx4QpZ6zhyl0jbxKfn5XB8xnCwmrajnTqvwRyqAXn+3DIG+HvmAdhY3a0","ud":"{\"a\":5}-","un":667}';
$renderer = new \BaconQrCode\Renderer\Image\Png();
$renderer->setHeight(210);
$renderer->setWidth(210);
$renderer->setMargin(0);
$writer = new \BaconQrCode\Writer($renderer);
$writer->writeFile($b, 'qrcode2.png', 'UTF-8', 2);
```

Results in the following images:

![qrcode1](https://cloud.githubusercontent.com/assets/3987203/25206156/3fc69752-2534-11e7-92e6-499a4cb64d53.png)
![qrcode2](https://cloud.githubusercontent.com/assets/3987203/25206159/434a652a-2534-11e7-90e0-aaf6a6329ff4.png)

Big padding is added to the second PNG, while first one is very close to the required 210x210 dimension.

Thanks!

**Comment by DASPRiD** [comment]
I just looked took a couple of minutes to look into the issue (I didn't touch this library in a while), which is actually rather complicated. To break it down: The renderer will add a quiet zone to make sure that each single block takes up a fixed amount of pixels. 

In your second QR code, there are 73 blocks in both directions. Given an image size of 210, which does not fit a block size of 3 pixels, it has to add padding so it can place blocks with a size of 2 pixels.

There is no proper way to work around this in the library. The best you can do, if you want consistent bitmap sizes and no quiet zone, is to render the image in a higher resolution (like, twice as big), then trim the image borders (Imagick can do that with a single method call) and then downscale it to 210 pixels, which would give you anti-aliased blocks.

If you have any further comments or ideas, feel free to re-open this issue.

**Comment by temuri416** [comment]
ok, thanks for looking into it.

do you know if there's a certain attribute of renderer class that would tell me that extra padding was applied?

**Comment by DASPRiD** [comment]
Nope, the renderers won't tell you that. But as I said, this is what you can do:

```php
// Create writer with size 420…

$image = new Imagick();
$image->readImageBlob($writer->writeString('…'));
$image->trimImage(0);
$image->resizeImage(210, 210, Imagick::FILTER_CATROM, 1);
$image->writeImage('qr-code.png');
```

**Comment by DASPRiD** [comment]
This should be fixed with the new PNG renderer in version 2.0.0, by the way.
